server/src/main/java/org/elasticsearch/action/support/AutoCreateIndex.java

```diff
@@ -67,6 +62,13 @@ public boolean shouldAutoCreate(String index, ClusterState state) {
         if (resolver.hasIndexAbstraction(index, state)) {
             return false;
         }
+
+        // Templates can override the AUTO_CREATE_INDEX_SETTING setting
+        final ComposableIndexTemplate template = findTemplate(index, state.metadata());
+        if (template != null && template.getAllowAutoCreate()) {
```

Comment:
We talked about the auto create index logic and the allow auto create template flag yesterday. This logic is invoked from the bulk action and is evaluated on the coordinating node. While the decision the auto create an index or data stream is done in the `AutoCreateAction` on the elected master node. 

The reason why we decided to make the decision to auto create an index or data stream in the `AutoCreateAction` and not here is, because if for some reason a node is lacking behind and doesn't have all the templates then an index is auto created instead of a data stream. Fixing the erroneous situation requires a lot of manual work, hence we moved making this decision to elected master node.

The question is whether the auto create index check (with this enhancement) should be moved to `AutoCreateAction` class too, so that all the decisions are made on the elected master node. 

I'm currently thinking the following:
* In the case a coordinating node doesn't have the latest `action.auto_create_index` setting values or composable index templates then the worst that can happen is a new few write requests fail, because the target index doesn't exist.
* Currently if `action.auto_create_index` is set to `false` then also data streams are no longer created. I think this is a bug, because the name of the setting implies auto creation of indices only. 
* If this logic is moved to `AutoCreateAction` class then the overhead is that remote requests are made to the elected master node in the case target indices don't exist. Most requests will index into an existing data stream or index, so the cost is acceptable.
* I remember that we eventually want to deprecate and remove the `action.auto_create_index` setting, but I may be mistaken.

If we are going to remove `action.auto_create_index` with the new allow auto create template flag then I think let's keep it here? And then maybe also add logic here that if a non existing index matches with a template that auto create data streams then checking the `action.auto_create_index` setting should be ignored?